include/yarpgen/type.h

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright (c) 2019, Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <limits>
+#include <climits>
+
+#include "ir_value.h"
+
+namespace yarpgen {
+
+// Abstract class that serves as a common ancestor for all types.
```

Comment:
Would be good to define what are we representing with Type - a target lanaguge type (i.e. int, uint32_t, intptr_t, etc) or backend type, which has certain properties and may be instatiated as different lanague types (i.e. signed 32 bit integer, which may be int, or int32_t).

Defining this may help avoiding confusion in the type system design.